src/workerd/api/gpu/gpu-buffer.c++

```diff
@@ -3,61 +3,181 @@
                   ? buffer_.GetMappedRange(o, s)
-                  : const_cast<void *>(buffer_.GetConstMappedRange(o, s));
+                  : const_cast<void*>(buffer_.GetConstMappedRange(o, s));
 
   JSG_REQUIRE(ptr, TypeError, "could not obtain mapped range");
 
-  auto arrayBuffer = kj::arrayPtr((byte *)ptr, s);
-  mapped_.add(Mapping{start, end, arrayBuffer});
+  struct Context {
+    jsg::Ref<GPUBuffer> buffer;
+  };
+  auto ref = JSG_THIS;
+  // We're creating this context object in order for it to be available when the
+  // callback is invoked. It owns a persistent reference to this GPUBuffer
+  // object to ensure that it still lives while the arraybuffer is in scope.
+  // This object will be deallocated when the callback finishes.
+  auto ctx = new Context{ref.addRef()};
+  std::shared_ptr<v8::BackingStore> backing = v8::ArrayBuffer::NewBackingStore(
+      ptr, s,
+      [](void* data, size_t length, void* deleter_data) {
+        // we have a persistent reference to GPUBuffer so that it lives at least
+        // as long as this arraybuffer
+        // Note: this is invoked outside the JS isolate lock
+        auto c = std::unique_ptr<Context>(static_cast<Context*>(deleter_data));
+      },
+      ctx);
+
+  v8::Local<v8::ArrayBuffer> arrayBuffer =
+      v8::ArrayBuffer::New(js.v8Isolate, backing);
+  arrayBuffer->SetDetachKey(detachKey_->getHandle(js.v8Isolate));
+
+  mapped_.add(
+      Mapping{start, end,
+              std::unique_ptr<jsg::V8Ref<v8::ArrayBuffer>>(
+                  new jsg::V8Ref<v8::ArrayBuffer>(js.v8Isolate, arrayBuffer))});
   return arrayBuffer;
 }
 
-void GPUBuffer::unmap() {
+void GPUBuffer::DetachMappings(jsg::Lock& js) {
+  for (auto& mapping : mapped_) {
+    auto ab = mapping.buffer->getHandle(js.v8Isolate);
+
+    auto res = ab->Detach(detachKey_->getHandle(js.v8Isolate));
+    KJ_ASSERT(res.IsJust());
+  }
+  mapped_.clear();
+}
+
+void GPUBuffer::destroy(jsg::Lock& js) {
+  if (state_ == State::Destroyed) {
+    return;
+  }
+
+  if (state_ != State::Unmapped) {
+    DetachMappings(js);
+  }
+
+  buffer_.Destroy();
+  state_ = State::Destroyed;
+}
+
+void GPUBuffer::unmap(jsg::Lock& js) {
   buffer_.Unmap();
 
   if (state_ != State::Destroyed && state_ != State::Unmapped) {
-    mapped_.clear();
+    DetachMappings(js);
     state_ = State::Unmapped;
   }
 }
 
+jsg::Promise<void> GPUBuffer::mapAsync(GPUFlagsConstant mode,
+                                       jsg::Optional<GPUSize64> offset,
+                                       jsg::Optional<GPUSize64> size) {
+  wgpu::MapMode md = static_cast<wgpu::MapMode>(mode);
+
+  if (state_ != State::Unmapped) {
+    device_.InjectError(
+        wgpu::ErrorType::Validation,
+        "mapAsync called on buffer that is not in the unmapped state");
+    JSG_FAIL_REQUIRE(
+        Error, "mapAsync called on buffer that is not in the unmapped state");
+  }
+
+  uint64_t o = offset.orDefault(0);
+  uint64_t s = size.orDefault(desc_.size - o);
```

Comment:
If this method is called multiple times without awaiting, it can queue up multiple requests for overlapping ranges. What's the expected behavior in that case? It would be good to at least add a comment here briefly explaining the expected behavior.

Reply:
We check the state of the buffer. Even if the promises are not `await`ed only the first one will succeed since the others will find the buffer with a state not equal to `State::Unmapped`.

Reply:
Added comment.